File: python-algorithm/leetcode/problem_350.py
"""350. Intersection of Two Arrays II
https://leetcode.com/problems/intersection-of-two-arrays-ii

Given two arrays, write a function to compute their intersection.

Example 1:

Input: nums1 = [1,2,2,1], nums2 = [2,2]
Output: [2,2]
Example 2:

Input: nums1 = [4,9,5], nums2 = [9,4,9,8,4]
Output: [4,9]
Note:

Each element in the result should appear as many times as it shows in both arrays.
The result can be in any order.
Follow up:

What if the given array is already sorted? How would you optimize your algorithm?
What if nums1's size is small compared to nums2's size? Which algorithm is better?
What if elements of nums2 are stored on disk, and the memory is limited
such that you cannot load all elements into the memory at once?
"""
from typing import List


class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        if len(nums1) > len(nums2):
            return self.intersect(nums2, nums1)
        store = {}
        for num in nums1:
            store[num] = store.get(num) + 1 if num in store else 1
        ans = []
        for num in nums2:
            if num in store and store.get(num) > 0:
                store[num] = store.get(num) - 1
                ans.append(num)
        return ans

    def intersect_1(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums1.sort()
        nums2.sort()
        i, j = 0, 0
        ans = []
        while i < len(nums1) and j < len(nums2):
            if nums1[i] == nums2[j]:
                ans.append(nums1[i])
                i += 1
                j += 1
            elif nums1[i] < nums2[j]:
                i += 1
            else:
                j += 1
        return ans

File: go-algorithm/leetcode/problem350.go
package leetcode

/*
350. Intersection of Two Arrays II
https://leetcode.com/problems/intersection-of-two-arrays-ii/
*/
func intersect(nums1 []int, nums2 []int) []int {
	var ans []int
	counter1, counter2 := make(map[int]int), make(map[int]int)
	for _, num := range nums1 {
		counter1[num]++
	}
	for _, num := range nums2 {
		counter2[num]++
	}
	for num, cnt := range counter1 {
		if counter2[num] != 0 {
			times := 0
			if cnt < counter2[num] {
				times = cnt
			} else {
				times = counter2[num]
			}
			for i := 0; i < times; i++ {
				ans = append(ans, num)
			}
		}
	}
	return ans
}
